File: app/src/main/java/com/backbase/citysearch/async/SearchCitiesTask.java
package com.backbase.citysearch.async;

import android.os.AsyncTask;

import com.backbase.citysearch.datastorage.CityDataStore;
import com.backbase.citysearch.models.City;
import java.util.List;

public class SearchCitiesTask extends AsyncTask<String, Void, List<City>> {
    private SearchCitiesTaskCallback callback;
    private CityDataStore cityDataStore;

    public SearchCitiesTask(SearchCitiesTaskCallback callback, CityDataStore cityDataStore) {
        this.callback = callback;
        this.cityDataStore = cityDataStore;
    }

    @Override
    protected List<City> doInBackground(String... strings) {
        String query = strings[0];
        return cityDataStore.findMatchingCities(query);
    }

    @Override
    protected void onPostExecute(List<City> cities) {
        super.onPostExecute(cities);
        callback.onSearchCompleted(cities);
    }

    public interface SearchCitiesTaskCallback {
        void onSearchCompleted(List<City> result);
    }
}

File: app/src/main/java/com/backbase/citysearch/async/SearchCitiesTask.kt
package com.backbase.citysearch.async

import android.os.AsyncTask
import com.backbase.citysearch.datastorage.CityDataStore
import com.backbase.citysearch.models.City

class SearchCitiesTask(private val callback: SearchCitiesTaskCallback, private val cityDataStore: CityDataStore) : AsyncTask<String?, Void?, List<City>?>() {

    override fun onPostExecute(cities: List<City>?) {
        super.onPostExecute(cities)
        callback.onSearchCompleted(cities)
    }

    interface SearchCitiesTaskCallback {
        fun onSearchCompleted(result: List<City>?)
    }

    override fun doInBackground(vararg params: String?): List<City>? {
        val query = params[0]
        return cityDataStore.findMatchingCities(query)
    }
}